### PR TITLE
[quant][graphmode] Support per channel quantization in insert_quant_dequant and fold_prepack

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29494 [quant][graphmode] Returning axis from calculate_qparams
* #29493 [quant][graphmode] Test for per channel graph mode quantization
* #29416 [quant] Make PerChannelMinMaxObserver scriptable using `torch.jit.ignore`
* #29701 [quant] Per channel quantization support in insert_prepack_unpack
* **#29492 [quant][graphmode] Support per channel quantization in insert_quant_dequant and fold_prepack**

Summary:
Previously graph mode quantization only works for per tensor quantization,
this PR added support for per channel quantization as well, changes include
- insert per channel quantization calls(insert_quant_dequant)
- add support of folding for prepacked per channel quantized weight (fold_prepack)

Test Plan:
test is not possible until we can script PerChannelObserver, which comes in https://github.com/pytorch/pytorch/pull/29416
we'll add test in a separate PR after that.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18580444](https://our.internmc.facebook.com/intern/diff/D18580444)

### CODE REVIEW

**Review by ZolotukhinM — APPROVED** [review]
Looks good modulo some small details.

File: torch/csrc/jit/passes/quantization.cpp [hunk 1]
```diff
@@ -391,13 +397,18 @@ Node* insertQuantDeQuantCall(
...
-  // Create qparam constant nodes
-  auto tp = qparams.toTuple();
-  IValue scale = tp->elements()[0].toTensor().item().toFloat();
-  IValue zero_point = tp->elements()[1].toTensor().item().toInt();
-  Value* scale_val = g->insertConstant(scale);
-  Value* zero_point_val = g->insertConstant(zero_point);
+  Value* scale_val;
+  Value* zero_point_val;
+  Value* axis_val;
+  if (is_per_channel) {
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: OK

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -997,14 +1033,21 @@ graph(%a_dequant, %w, %b, %w_scale, %w_zero_point, %w_dtype, %stride, %padding,
         matches.size() <= 1, "We only support at most one match right now");
     for (const auto& match : matches) {
       const auto& match_vmap = match.values_map;
-      auto w_scale =
-          toIValue(match_vmap.at(vmap.at("w_scale"))).value().toDouble();
-      auto w_zero_point =
-          toIValue(match_vmap.at(vmap.at("w_zero_point"))).value().toInt();
-      auto w_dtype =
-          toIValue(match_vmap.at(vmap.at("w_dtype"))).value().toScalarType();
+      auto w_dtype = getIValue("w_dtype", match_vmap, vmap).toScalarType();
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: then there is not much value putting this into a function.. it only saves a little bit of typing anyways

[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: we are assuming they are IValue, if not, this match should fail. I can add a check to skip the ones that fails

**Comment by jerryzh168** [comment]
@ZolotukhinM can you take a look again, I did a relatively big change

File: torch/csrc/jit/passes/quantization.cpp [hunk 3]
```diff
@@ -376,47 +373,40 @@ void InsertObserversHelper::insertObservers(
...
+  if (is_per_channel) {
+    quantize_func ="quantize_per_channel";
+    inputs.push_back(g->insertConstant(scale));
+    inputs.push_back(g->insertConstant(zero_point));
+    inputs.push_back(g->insertConstant(0));
   } else {
-    quant->insertBefore(insert_point);
+    quantize_func = "quantize_per_tensor";
+    inputs.push_back(g->insertConstant(scale.item<double>()));
+    inputs.push_back(g->insertConstant(zero_point.item<long>()));
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: This breaks all the builds:
```
../torch/csrc/jit/passes/quantization.cpp:396:40: error: conversion from 'long' to 'c10::IValue' is ambiguous
     inputs.push_back(g->insertConstant(zero_point.item<long>()));
                                        ^~~~~~~~~~~~~~~~~~~~~~~
```

File: torch/csrc/jit/passes/quantization.cpp [hunk 4]
```diff
@@ -962,76 +954,109 @@ void InsertPrepackUnpack(script::Module& module) {
   }
 }
 
+static c10::optional<IValue> getIValue(const std::string& name,
+                                       const std::unordered_map<const Value*, Value*>& match_vmap,
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Looks like it needs clang-format.

File: torch/csrc/jit/passes/quantization.cpp [hunk 5]
```diff
@@ -962,76 +954,109 @@ void InsertPrepackUnpack(script::Module& module) {
...
-          toIValue(match_vmap.at(vmap.at("w_scale"))).value().toDouble();
-      auto w_zero_point =
-          toIValue(match_vmap.at(vmap.at("w_zero_point"))).value().toInt();
-      auto w_dtype =
-          toIValue(match_vmap.at(vmap.at("w_dtype"))).value().toScalarType();
+      auto w_dtype_opt = getIValue("w_dtype", match_vmap, vmap);
+      auto w_scale_opt = getIValue("w_scale", match_vmap, vmap);
+      auto w_zero_point_opt = getIValue("w_zero_point", match_vmap, vmap);
+      if (!w_dtype_opt || !w_scale_opt || !w_zero_point_opt) {
+        GRAPH_DEBUG("dtype, scale or zero_point for weight is not constant, skipping the match.");
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Thanks for adding the logging! To make it more helpful it's better to add more details to the message: e.g. something to locate these values in the actual graph. Also, usually we print the graph before or after transformation with `GRAPH_DUMP` and report mutations we did on the graph with `GRAPH_UPDATE`.

**Review by ZolotukhinM — APPROVED** [review]
This looks good! Please find some comments inline and fix the compilation error :)

**Pull request closed without merging**